core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -407,11 +407,10 @@ private ThreadContextStruct putHeaders(Map<String, String> headers) {
             if (headers.isEmpty()) {
                 return this;
             } else {
-                final Map<String, String> newHeaders = new HashMap<>();
+                final Map<String, String> newHeaders = new HashMap<>(this.requestHeaders);
```

Comment:
I think you misunderstood my comment. Look a the implementation of `putSingleHeader`, which is what the loop over `headers` calls to add each header. It will throw an exception if a key already exists. This means with your change if a key exists in `requestHeaders` and `headers`, an exception will be thrown, while before, the value from `requestHeaders` would have one. The bug here is that the `requestHeaders` value should be the fallback, right? So then, instead of initializing the `newHeaders` to `requestHeaders`, you can iterate over `requestHeaders` in place of the current `putAll` call, and use `newHeaders.putIfAbsent`, so that the value would only be added if `headers` did not contain that key.